pkg/controller/ipam/ipam_test.go

```diff
@@ -0,0 +1,421 @@
+			nil,
+			makeNode("bootstrap-node", map[string]string{"bootstrap": "true"}),
+			[]string{"10.2.1.0/24", "fd00:20:96::100/120"},
+		),
+		ginkgo.Entry("Single stack IPv4 Pod CIDR assigned to a node",
+			makeClusterCIDR("ipv4-cc", "10.0.0.0/16", "", 8, nodeSelector(map[string][]string{"ipv4": {"true"}, "singlestack": {"true"}})),
+			makeNode("ipv4-node", map[string]string{"ipv4": "true", "singlestack": "true"}),
+			[]string{"10.0.0.0/24"},
+		),
+		ginkgo.Entry("Single stack IPv6 Pod CIDR assigned to a node",
+			makeClusterCIDR("ipv6-cc", "", "fd00:20:100::/112", 8, nodeSelector(map[string][]string{"ipv6": {"true"}})),
+			makeNode("ipv6-node", map[string]string{"ipv6": "true"}),
+			[]string{"fd00:20:100::/120"},
+		),
+		ginkgo.Entry("DualStack Pod CIDRs assigned to a node",
+			makeClusterCIDR("dualstack-allocate-cc", "192.168.0.0/16", "fd00:30:100::/112", 8, nodeSelector(map[string][]string{"ipv4": {"true"}, "ipv6": {"true"}})),
+			makeNode("dualstack-allocate-node", map[string]string{"ipv4": "true", "ipv6": "true"}),
+			[]string{"192.168.0.0/24", "fd00:30:100::/120"},
+		),
+	)
+
+	ginkgo.It("should release Pod CIDR after node is deleted", func() {
+		// Create the test ClusterCIDR.
+		clusterCIDR := makeClusterCIDR("dualstack-release-cc", "192.168.0.0/23", "fd00:30:100::/119", 8, nodeSelector(map[string][]string{"ipv4": {"true"}, "ipv6": {"true"}}))
+		_, err := cidrClient.NetworkingV1().ClusterCIDRs().Create(ctx, clusterCIDR, metav1.CreateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		clusterCIDRsToCleanup = []*v1.ClusterCIDR{clusterCIDR}
+
+		// Create 1st node and validate that Pod CIDRs are correctly assigned.
+		node1 := makeNode("dualstack-release-node", map[string]string{"ipv4": "true", "ipv6": "true"})
+		expectedPodCIDRs1 := []string{"192.168.0.0/24", "fd00:30:100::/120"}
+		_, err = kubeClient.CoreV1().Nodes().Create(ctx, node1, metav1.CreateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		gomega.Eventually(komega.Object(node1)).Should(gomega.WithTransform(func(n *corev1.Node) []string {
+			return n.Spec.PodCIDRs
+		}, gomega.Equal(expectedPodCIDRs1)))
+
+		// Create 2nd node and validate that Pod CIDRs are correctly assigned.
+		node2 := makeNode("dualstack-release-node-2", map[string]string{"ipv4": "true", "ipv6": "true"})
+		expectedPodCIDRs2 := []string{"192.168.1.0/24", "fd00:30:100::100/120"}
+		_, err = kubeClient.CoreV1().Nodes().Create(ctx, node2, metav1.CreateOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		nodesToCleanup = append(nodesToCleanup, node2)
+
+		gomega.Eventually(komega.Object(node2)).Should(gomega.WithTransform(func(n *corev1.Node) []string {
+			return n.Spec.PodCIDRs
+		}, gomega.Equal(expectedPodCIDRs2)))
+
+		// Delete the 1st node, to validate that the PodCIDRs are released.
+		gomega.Expect(kubeClient.CoreV1().Nodes().Delete(ctx, node1.Name, metav1.DeleteOptions{})).To(gomega.Succeed())
+
```

Comment:
Adding a delay here might be helpful, I am worried it could lead to a flaky test if the new node is created before the CIDR is released

Reply:
added